formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt

```diff
@@ -0,0 +1,636 @@
+    ): Int {
+        processTags(tags)
+        if (curByte == unboundedHeader) {
+            skipByte(unboundedHeader)
+            return -1
+        }
+        if ((curByte and 0b111_00000) != boundedHeaderMask)
+            throw CborDecodingException("start of $collectionType", curByte)
+        val size = readNumber().toInt()
+        readByte()
+        return size
+    }
+
+    fun isEnd() = curByte == BREAK
+
+    fun end() = skipByte(BREAK)
+
+    fun nextByteString(tags: ULongArray? = null): ByteArray {
+        processTags(tags)
+        if ((curByte and 0b111_00000) != HEADER_BYTE_STRING)
+            throw CborDecodingException("start of byte string", curByte)
+        val arr = readBytes()
+        readByte()
+        return arr
+    }
+
+    fun nextString(tag: ULong) = nextString(ulongArrayOf(tag))
+    fun nextString(tags: ULongArray? = null) = nextTaggedString(tags).first
+
+    //used for reading the tag names and names of tagged keys (of maps, and serialized classes)
+    private fun nextTaggedString(tags: ULongArray?): Pair<String, ULongArray?> {
+        val collectedTags = processTags(tags)
+        if ((curByte and 0b111_00000) != HEADER_STRING)
+            throw CborDecodingException("start of string", curByte)
+        val arr = readBytes()
+        val ans = arr.decodeToString()
+        readByte()
+        return ans to collectedTags
+    }
+
+    private fun readBytes(): ByteArray =
+        if (curByte and 0b000_11111 == ADDITIONAL_INFORMATION_INDEFINITE_LENGTH) {
+            readByte()
+            readIndefiniteLengthBytes()
+        } else {
+            val strLen = readNumber().toInt()
+            input.readExactNBytes(strLen)
+        }
+
+    @SuppressAnimalSniffer
+    private fun processTags(tags: ULongArray?): ULongArray? {
+        var index = 0
+        val collectedTags = mutableListOf<ULong>()
+        while ((curByte and 0b111_00000) == HEADER_TAG) {
+            val readTag = readNumber().toULong() // This is the tag number
+            collectedTags += readTag
+            // value tags and object tags are intermingled (keyTags are always separate)
+            // so this check only holds if we verify both
+            if (verifyObjectTags) {
+                tags?.let {
+                    if (index++ > it.size) throw CborDecodingException("More tags found than the ${it.size} tags specified")
+                    if (readTag != it[index - 1]) throw CborDecodingException("CBOR tag $readTag does not match expected tag $it")
+                }
+            }
+            readByte()
+        }
+        return (if (collectedTags.isEmpty()) null else collectedTags.toULongArray()).also { collected ->
+            //We only want to compare if tags are actually set, otherwise, we don't care
+            tags?.let {
+                if (verifyObjectTags) { //again, this check only works if we verify value tags and object tags
+                    if (!it.contentEquals(collected))
```

Comment:
Didn't we check the equality of tags in the previous `while`? What's the difference here?

Reply:
Wow! Awesome catch!  I really though I had all mistagged cases covered. that one is fixed and has a test case for it too in 6cd632416cabe12e15c6094e6e33cf3b78b9d262.

Reply:
So I now officially don't know what the GitHub website is doing. Everything is Jumping around when all conversations are shown and apparently this comment got replaces by an entirely different one. Maybe some of my comments are now placced where the don't make sense. Anyways, I removed the second check in the while loop, because, it can never be triggered. The while loop causes an early exit, when more tags are found.
The while loop needs to be carried out anyhow, to consume all tags. once this is done, the actual comparison starts. The check for verifyObjectTags is necessary, because only then a complete check is possible, otherwise we can only check whether the parsed tags start with the expected value tags (this is the else branch).